### PR TITLE
Deprecate manual webhook key input and introduce automatic webhook configuration

### PR DESCRIPTION
**Pull request by james-allan**
Part of #3195 

## Changes proposed in this Pull Request:

This PR replaces the old Webhook key input for button that can be used to automatically configure Webhooks for you. 

| Not configured | Configured |
|--------|--------|
| Screenshot 2024-06-18 at 1 01 04 PM | Screenshot 2024-06-18 at 1 00 02 PM | 

### Previously entered webhook secret

1. On `develop` go to your Stripe API settings and enter a webhook secret. 
2. Checkout this branch.
3. run `npm run build`
4. Refresh the Stripe plugin settings page and open the api keys modal.
5. You should see the "Reconfigure webhooks" button. Notice that no URL is shown as we don't have any information that webhook. We only have the secret.

<p align="center">
Screenshot 2024-06-18 at 3 00 46 PM
</p>

### No webhook secret

1. Run the following code snippet to delete your webhook secrets. 

```php
$settings = get_option( 'woocommerce_stripe_settings', [] );

unset($settings['webhook_secret'] );
unset($settings['test_webhook_secret'] );

update_option( 'woocommerce_stripe_settings', $settings );
```

2. Refresh the Stripe settings page and open the API keys modal. 
3. You should now see the option to "Configure webhooks" button. 

<p align="center">
Screenshot 2024-06-18 at 3 05 27 PM
</p>

4. Click the **Configure webhooks** button.
5. Once the request is handled you should now see the **Enabled** status, the button would have changed to "**Recongifure webhooks**" and information about where the webhooks will be sent is shown.
6. Clicking **Recongifure webhooks** will delete the old webhook and create a new one. You can verify this by checking your `woocommerce_stripe_settings` comparing the webhook secret or ID with the one in your Stripe Dashboard. 

### Some other things you may want to confirm/test

1. Entering a new secret key will use that secret key to configure the webhooks rather than the one stored in settings.
2. Emptying out the secret key field will disable the configure button.
3. While saving API keys, the configure button is disabled. 

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

File: client/settings/account-details/index.js [hunk 1]
```diff
@@ -5,8 +5,8 @@ import { Button, ExternalLink, Icon } from '@wordpress/components';
 import interpolateComponents from 'interpolate-components';
 import { help } from '@wordpress/icons';
 import styled from '@emotion/styled';
+import SectionStatus from '../section-status';
 import useWebhookStateMessage from './use-webhook-state-message';
-import SectionStatus from './section-status';
```
Comment (james-allan) [inline client/settings/account-details/index.js:9]: I moved the `section-status` component so it was more accessible. Its now in the `client/settings` hence why this change was necessary. :)

File: client/settings/payment-settings/account-keys-modal.js [hunk 2]
```diff
@@ -73,7 +79,10 @@ const SecretKey = () => {
 				'woocommerce-gateway-stripe'
 			) }
 			value={ value }
-			onChange={ ( val ) => setValue( val ) }
+			onChange={ ( val ) => {
+				setValue( val );
+				setSecretKey( val );
```
Comment (james-allan) [inline client/settings/payment-settings/account-keys-modal.js:84]: When the user changes their secret, I now call the setter (`setSecretKey`) so other consumers of the `useAccountKeysSecretKey()` hook can get the latest key.

**Comment by james-allan** [comment]
I'm submitting this for review now, however, I wanted to highlight some other changes I plan here. 

1. I think it would be a good idea to rate limit this request to configure webhooks so users don't click the button rapidly and send off multiple requests to Stripe. I'll work on this separately as adding a rate limit to this PR would make testing more complicated. 
2. There's an issue where if you remove your API keys, the webhook remains enabled. I think I want to look into better handling this. ie 1 solution may be to also delete the webhook when you disconnect your store. 

I've filed those as separate tasks on the epic task list ([here](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3195)).

**Comment by james-allan** [comment]
@aheckler just a friendly FYI. This change is planned for Stripe 8.5 (July 11). 

Given this changes the admin API credentials UI I wanted to make sure I flagged it with you. It hasn't been reviewed yet and so things may change before it's eventually merged.

File: client/data/account-keys/actions.js [hunk 3]
```diff
@@ -111,3 +111,44 @@ export function* testAccountKeys( { live, publishable, secret } ) {
...
 }
+
+export function updateIsConfiguringWebhooks( isProcessing ) {
+	return {
+		type: ACTION_TYPES.SET_IS_CONFIGURING_WEBHOOKS,
+		isProcessing,
+	};
+}
+
+export function* configureWebhooks( { live, secret, callback } ) {
```
Comment (a-danae) [inline client/data/account-keys/actions.js:122]: This works great; I'm just curious. Is there any difference between using this as a generator function and using async/await?

I see we're using generators in other places too and this works well, so not asking to update it.

File: includes/admin/class-wc-rest-stripe-account-keys-controller.php [hunk 4]
```diff
@@ -335,6 +357,94 @@ public function test_account_keys( WP_REST_Request $request ) {
...
+	 */
+	public function configure_webhooks( WP_REST_Request $request ) {
+		$settings     = get_option( self::STRIPE_GATEWAY_SETTINGS_OPTION_NAME, [] );
+		$live_mode    = wc_clean( wp_unslash( $request->get_param( 'live_mode' ) ) );
+		$secret       = wc_clean( wp_unslash( $request->get_param( 'secret' ) ) );
+		$saved_secret = $settings[ $live_mode ? 'secret_key' : 'test_secret_key' ];
+
+		// Check if the user is configuring the opposite mode. ie if the store is in live mode and is configuring webhooks for test mode.
+		$is_testmode_enabled     = 'yes' === $settings['testmode'];
+		$configure_opposite_mode = isset( $settings['testmode'] ) && ( $live_mode && $is_testmode_enabled ) || ( ! $live_mode && ! $is_testmode_enabled );
```
Comment (a-danae) [inline includes/admin/class-wc-rest-stripe-account-keys-controller.php]: Would the `isset( $settings['testmode'] )` check be needed here if we're doing `$is_testmode_enabled = 'yes' === $settings['testmode'];` in the line above?

Also, wouldn't it be an error if `$settings['testmode']` isn't set before reaching this line?

**Review by a-danae — APPROVED** [review]
Nice changes here, @james-allan ! 

✅ The webhooks are configured as expected when using this new flow
✅ The behaviors mentioned in the PR details work as described
✅ The code looks good and makes sense to me

❓ When using a secret key from a different account, the webhook gets configured in this new account but the store continues to be connected to the old one. Should we handle this differently? We could save the keys before configuring the webhook, for example.

This is an edgy scenario, and it can be handled separately if we want to adjust it.

Steps:
1. Connect your store to an account
2. Change the secret key to one of a new account but don't save it
3. Click on configure webhook
4. Close the modal without saving
5. Notice that the webhook is configured on the new account, but the key isn’t saved. So the store is connected to an account, and the webhook is configured in another.

File: client/data/account-keys/actions.js (hunk 3 above)
Comment (james-allan) [reply client/data/account-keys/actions.js:122]: Good question. I'm not 100% sure 😅 

I copied the precedence set by the other function in this file. TBH I wasn't really sure what the pros/cons were of using a generator function vs async but ChatGPT did say async is generally more preferred.

File: includes/admin/class-wc-rest-stripe-account-keys-controller.php (hunk 4 above)
Comment (james-allan) [reply includes/admin/class-wc-rest-stripe-account-keys-controller.php]: Yeah good point. I've rejigged this [41267c3](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3209/commits/41267c37756cb32aa5672c3e528eb29c5ac06ce0). I now check if test mode is set in the `$is_test_mode_enabled` definition and then removed it from the next line.

**Comment by james-allan** [comment]
> This is an edgy scenario, and it can be handled separately if we want to adjust it.

Yeah I think we can handle that separately. It sort of has similarities to https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3221. Perhaps I can include it there.

File: client/data/account-keys/actions.js (hunk 3 above)
Comment (james-allan) [reply client/data/account-keys/actions.js:122]: I'm going to leave this as a generator function for now. We can change it to async in a separate change if we prefer that approach. I'd prefer to keep this PR moving. :)

**Comment by james-allan** [comment]
> Yeah I think we can handle that separately. It sort of has similarities to https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3221. Perhaps I can include it there.

I just realised it's the opposite to the problem being fixed in #3221 and so it probably makes sense to fix it in a new PR. 

I'm not sure about saving the secret key when you click the configure button. I think we'd need to communicate that better. eg if the user enters keys, the button text should change to "Save keys and configure webhooks" or at least flag that in this case the keys will also be saved - perhaps with a note.

**Comment by a-danae** [comment]
> I think we'd need to communicate that better. eg if the user enters keys, the button text should change to "Save keys and configure webhooks" or at least flag that in this case the keys will also be saved - perhaps with a note.

That makes sense. The fact that the keys would be saved in this scenario would need to be more intuitive for the merchant.

Would this still be relevant after the changes introduced in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3269 ? I can create a separate issue for this if so.

**Comment by james-allan** [comment]
> Would this still be relevant after the changes introduced in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3269 ? I can create a separate issue for this if so.

Nah I don't think it will be relevant after we deprecate the api key fields. That's largely why I decided not to pursue anything further around this.
